### PR TITLE
SSR improvements

### PR DESCRIPTION
**Pull request by antho1404**
Dependency: https://github.com/liteflow-labs/starter-kit/pull/241

### Description

Update SSR and use `getDataTree` to resolve all graphql queries
Improve the `useAccount` to have better compatibility on the server side
Fix cache issue with collection composed key

No more `getServerSideProps` 🎉 

### How to test

You can just run the different website pages with and without a refresh and check that the queries to get the data in the network panel do not have the pages' queries.
Also, you can use the settings `disable javascript` to really see the ssr.

### Checklist

- [x] Base branch of the PR is `dev`
- [ ] Update docs if necessary

### CODE REVIEW

File: client.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    uri: environment.GRAPHQL_URL,
+    headers: authorization ? { authorization: `Bearer ${authorization}` } : {},
+    cache: new InMemoryCache({
+      typePolicies: {
+        Account: {
+          keyFields: ['address'],
+        },
+        Collection: {
+          keyFields: ['chainId', 'address'],
+        },
```
[the remark this answers is not in the document]
Comment (NicolasMahe) [reply client.ts:29]: is it time to active the `nodeId` on the GraphQL schema?

File: hooks/useRequiredQueryParamSingle.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    parse?: (value: string) => T
+  } = {},
+): T {
+  const value = useQueryParamSingle<T>(key, {
+    parse(value) {
+      invariant(value)
+      return options.parse ? options.parse(value) : (value as T)
+    },
+  })
+  invariant(value)
```
Comment (NicolasMahe) [inline hooks/useRequiredQueryParamSingle.ts:16]: that's going to throw 500 errors when the url is malformed. should we use the `useQueryParamSingle` version and manage this in the pages?

File: pages/checkout/[id].tsx [hunk 3]
```diff
@@ -28,80 +26,25 @@ import {
...
-
-const CheckoutPage: NextPage<Props> = ({
-  now,
-  offerId,
-  meta,
-  currentAccount,
-}) => {
-  const ready = useEagerConnect()
+const CheckoutPage: NextPage<Props> = ({ now }) => {
+  useEagerConnect()
```
Comment (NicolasMahe) [inline pages/checkout/[id].tsx:42]: Seems `useEagerConnect` is not doing anything. Could be removed.
```suggestion
```

Comment (NicolasMahe) [reply pages/checkout/[id].tsx:42]: Same comment for other pages using `useEagerConnect`

File: pages/create/index.tsx [hunk 4]
```diff
@@ -75,20 +49,20 @@ const Layout = ({ children }: { children: React.ReactNode }) => (
...
         account: address || '',
         collectionFilter: collectionsFilter,
         fetchCollections: environment.MINTABLE_COLLECTIONS.length > 0,
       },
-      skip: !ready,
-    },
-  )
+    })
+
+  if (loading) return <></>
```
Comment (NicolasMahe) [inline pages/create/index.tsx:65]: Why adding this special render on loading here? Is it breaking the render without it?

File: pages/_app.tsx [hunk 5]
```diff
@@ -1,17 +1,13 @@
-import {
```
Comment (NicolasMahe) [inline pages/_app.tsx:1]: Is there a simple way to block the page from changing until it's content is loaded?

On almost every change of page, the website is blank until the queries are resolved.

https://user-images.githubusercontent.com/5823445/229420725-45ac2cc1-e30f-410d-a6c3-cda73254b751.mov

The experience of navigating the website is now pretty terrible because of this.

Only the page users bid, trades and offers seems to display placeholder data until its query are resolved, but they are missing a loader: https://nft-test-bsc-testnet-git-feature-ssr-improvement-liteflow.vercel.app/users/0x4b595014f7b45789c3f4e79324ae6d8090a6c8b5/bids

File: pages/tokens/[id]/bid.tsx [hunk 6]
```diff
@@ -30,111 +28,34 @@ import {
...
-        },
-        currentAccount: ctx.user.address,
-      },
-    }
-  },
-)
-
-const BidPage: NextPage<Props> = ({ now, assetId, meta, currentAccount }) => {
-  const ready = useEagerConnect()
+const BidPage: NextPage<Props> = ({ now }) => {
```
Comment (NicolasMahe) [inline pages/tokens/[id]/bid.tsx:44]: There is an issue when I navigate to this page for the first time from another page.
It's working well after I refresh the page.

https://user-images.githubusercontent.com/5823445/229422462-9ac3717b-3e0b-4f14-9470-1613699aadc0.mov

**Review by NicolasMahe — COMMENTED** [review]
That's much simpler change than I thought!
But there is a lot of work to do to make the experience as good as before, improving the loading state, better cache management and invalidation, etc...

File: client.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (antho1404) [reply client.ts:29]: Might help indeed but let's not do that right now

File: pages/checkout/[id].tsx (hunk 3 above)
Comment (antho1404) [reply pages/checkout/[id].tsx:42]: Yes might be, but it's tracked here https://github.com/liteflow-labs/starter-kit/issues/231 so I didn't want to push on that as this requires dedicated testing which is not the goal of this PR

File: pages/create/index.tsx (hunk 4 above)
Comment (antho1404) [reply pages/create/index.tsx:65]: Yes if we don't, then data are not present and then we render the wrong information and we have flickering between wrong and right information which is worse than no information and information

File: pages/_app.tsx (hunk 5 above)
Comment (antho1404) [reply pages/_app.tsx:1]: No, the only way is to properly handle loading states and/or to break the queries into smaller pieces (which is my goal). Still with loaders when required. I will work on loader on another PR(s) and merge into this one

File: pages/tokens/[id]/bid.tsx (hunk 6 above)
Comment (antho1404) [reply pages/tokens/[id]/bid.tsx:44]: fixed in this PR as it is not really related to the SSR but a more general problem with forms that was highlighted here https://github.com/liteflow-labs/starter-kit/pull/241

File: pages/_app.tsx [hunk 7]
```diff
@@ -287,4 +262,24 @@ function MyApp({
...
+
+MyApp.getInitialProps = async (
+  appContext: AppContext,
+): Promise<AppInitialProps<MyAppProps>> => {
+  const initialProps = await App.getInitialProps(appContext)
+  const cookies = (appContext.ctx.req as any)?.cookies // TODO: fix type
+  const jwt = cookies ? cookies['jwt-token'] : null
+  // Generate the now time, rounded to the second to avoid re-rendering on the client
+  // TOFIX: find a better way to share the time between the app and document
+  const now = Math.floor(Date.now() / 1000) * 1000
```
[the remark this answers is not in the document]
Comment (NicolasMahe) [reply pages/_app.tsx:274]: Did you try to inject the variable `now` in the document's `__NEXT_DATA__` like the `apolloState`?
It may be possible to access this variable from the app if it inside the `__NEXT_DATA__` in the right place.

File: pages/_document.tsx [hunk 8]
```diff
@@ -19,6 +38,27 @@ class MyDocument extends Document {
       </Html>
     )
   }
+
+  static async getInitialProps(
+    context: DocumentContext,
+  ): Promise<DocumentInitialProps & { apolloState: NormalizedCacheObject }> {
+    invariant(context.req)
+    const cookies = (context.req as any).cookies // TODO: fix type
```
Comment (NicolasMahe) [inline pages/_document.tsx:46]: The `TODO: fix type` have been already fixed on the admin: https://github.com/liteflow-labs/admin/pull/82
Do you want to do it in another PR?

**Review by NicolasMahe — APPROVED** [review]
It's working pretty good but we now need to focus on improving the loading/flickering and invalidation of local cache.

File: pages/_app.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (antho1404) [reply pages/_app.tsx:274]: Yes, the issue here is that this is called twice on the server side, so not yet injected into the document.

[the remark this answers is not in the document]
Comment (antho1404) [reply pages/_app.tsx:274]: https://github.com/liteflow-labs/starter-kit/issues/250

File: pages/_document.tsx (hunk 8 above)
Comment (antho1404) [reply pages/_document.tsx:46]: Yes I will do that in another PR to not block this one and add another review process
